### ISSUE TITLE
TRPO或者PPO程序疑问

### ISSUE THREAD

**Issue by lz200202** [boyu-ai/Hands-on-RL#96]
![image](https://github.com/user-attachments/assets/d55ab88c-1997-41f4-8dee-51a730f5a037)
old_log_probs计算完成后立即开始计算log_probs，它们之间并没有进行策略梯度参数更新的操作，为什么他们不相等？

**Comment by YanikZ35** [comment]
你如果用了==这样的判据的话确实不是相等的，因为前者detach了而后者没有，可以像代码里这个方法中一样，看一下两个的ratio是不是1
    def compute_surrogate_obj(self, states, actions, advantage, old_log_probs, actor):  # 计算策略目标
        log_probs = torch.log(actor(states).gather(1, actions))
        ratio = torch.exp(log_probs - old_log_probs)
        return torch.mean(ratio * advantage)

**Comment by xumanba** [comment]
他们在第一次的时候是相等的，后面log-probs会被梯度更新，但是old probs不会变，保证模型在几个epoch的训练下是不变的

**Comment by sunshinesd** [comment]
> 他们在第一次的时候是相等的，后面log-probs会被梯度更新，但是old probs不会变，保证模型在几个epoch的训练下是不变的

第一次相等，导致actor梯度为0 ，策略网络参数不会更新，进入第二次循环，log-probs还会跟之前一样，又会相等，梯度又会为0 ，不是进入一个不会变化的死循环吗？critic网络虽然有更新，但并没有在循环中重新计算A，所以A的结果也不会有变。对actor不会有影响

**Comment by sunshinesd** [comment]
> 他们在第一次的时候是相等的，后面log-probs会被梯度更新，但是old probs不会变，保证模型在几个epoch的训练下是不变的

运行代码，确实第一次相等，ratio=1，但actor网络的梯度不为0，导致后面log-probs会更新，整个系统就启动了。但问题是，ratio=1的时候，actor网络的梯度应该=0才对啊，为什么运行后梯度不等于0呢？

**Comment by sunshinesd** [comment]
> 他们在第一次的时候是相等的，后面log-probs会被梯度更新，但是old probs不会变，保证模型在几个epoch的训练下是不变的

运行代码，确实第一次相等，ratio=1，但actor网络的梯度不为0，导致后面log-probs会更新，整个系统就启动了。但问题是，ratio=1的时候，actor网络的梯度应该=0才对啊，为什么运行后梯度不等于0呢？

**Comment by sunshinesd** [comment]
> 他们在第一次的时候是相等的，后面log-probs会被梯度更新，但是old probs不会变，保证模型在几个epoch的训练下是不变的

运行代码，确实第一次相等，ratio=1，但actor网络的梯度不为0，导致后面log-probs会更新，整个系统就启动了。但问题是，ratio=1的时候，整个目标函数只跟A有关，跟actor网络参数无关，actor网络的梯度应该=0才对啊，为什么运行后梯度不等于0呢？

**Comment by sunshinesd** [comment]
> 他们在第一次的时候是相等的，后面log-probs会被梯度更新，但是old probs不会变，保证模型在几个epoch的训练下是不变的

运行代码，确实第一次相等，ratio=1，但actor网络的梯度不为0，导致后面log-probs会更新，整个系统就启动了。但问题是，ratio=1的时候，整个目标函数只跟A有关，跟actor网络参数无关，actor网络的梯度应该=0才对啊，为什么运行后梯度不等于0呢？

**Comment by sunshinesd** [comment]
> > 他们在第一次的时候是相等的，后面log-probs会被梯度更新，但是old probs不会变，保证模型在几个epoch的训练下是不变的
> 
> 运行代码，确实第一次相等，ratio=1，但actor网络的梯度不为0，导致后面log-probs会更新，整个系统就启动了。但问题是，ratio=1的时候，整个目标函数只跟A有关，跟actor网络参数无关，actor网络的梯度应该=0才对啊，为什么运行后梯度不等于0呢？

搞清楚了， 第一次更新时，log_prob和old_log_prob相等，但梯度不为零，原因在于log_prob和old_log_prob只是数值相等，不代表跟动作输出概率无关，也就不代表跟策略参数无关。事实上，由于old_log_prob的detach()操作，old_log_prob可以被看成与变量无关的常数，而log_prob才是真正的变量。首次更新时，只能说log_prob变量的值正好跟old_log_prob相等，不代表函数跟变量log_prob无关。

**Comment by zhfuch** [comment]
> 你如果用了==这样的判据的话确实不是相等的，因为前者detach了而后者没有，可以像代码里这个方法中一样，看一下两个的ratio是不是1 def compute_surrogate_obj(self, states, actions, advantage, old_log_probs, actor): # 计算策略目标 log_probs = torch.log(actor(states).gather(1, actions)) ratio = torch.exp(log_probs - old_log_probs) return torch.mean(ratio * advantage)

可是看上边的代码old_log_probs和log_probs 是放在一起计算的，即使后边迭代了计算出来的不应该都一样吗？因为他们俩是用同一个actor计算的啊，很困惑

**Comment by shihaofu2003** [comment]
> 他们在第一次的时候是相等的，后面log-probs会被梯度更新，但是old probs不会变，保证模型在几个epoch的训练下是不变的

你好，请问“保证模型在几个epoch的训练下是不变的”想表达的意思是否是： 如果在每个 episode 下进行多次 Policy Model 的更新，固定 $\pi_{old}$, 这样只有第一次更新时 $log\\_prob\\_old == log\\_prob\\_new$, 从第二次更新开始不再相等。

书中代码实现的逻辑是 **一个 episode 下更新一次 Policy**，也就意味着每一次更新时 $ratio = 1$。

**Comment by xumanba** [comment]
@shihaofu2003 是这个意思
